### PR TITLE
Multithread

### PR DESCRIPTION
**Pull request by mhrmm**
@jkinkead Probably you'll have something to say about this one. Not sure I'm doing multithreading in an AI2-approved way.

### CODE REVIEW

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/RandomForest.scala [hunk 1]
```diff
@@ -100,7 +101,13 @@ class RandomForestTrainer(validationPercentage: Double, numDecisionTrees: Int,
     * @return the induced random forest
     */
   override def apply(data: FeatureVectorSource): ProbabilisticClassifier = {
-    val subtreeFiles = Seq.fill(numDecisionTrees) {
+    import scala.concurrent.duration._
+    import scala.concurrent._
+    import scala.language.postfixOps
+    implicit val executionContext = ExecutionContext.fromExecutor(
```
Comment (jkinkead) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/RandomForest.scala]: You're probably better off just using the default execution context.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/ParseFile.scala [hunk 2]
```diff
@@ -66,11 +62,20 @@ object ParseFile {
   def parseTestSet(parser: TransitionParser, parseSource: PolytreeParseSource): Unit = {
     println("Parsing test set.")
     val startTime: Long = Platform.currentTime
-    val candidateParses: Iterator[Option[PolytreeParse]] = {
-      parseSource.parseIterator map {
-        parse => parser.parse(parse.sentence)
+
+    import scala.concurrent.duration._
```
Comment (jkinkead) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/ParseFile.scala]: You should move imports to the top of the file.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/ParseFile.scala [hunk 3]
```diff
@@ -66,11 +62,20 @@ object ParseFile {
...
+    import scala.concurrent._
```
Comment (jkinkead) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/ParseFile.scala]: I think you're just using `{ Await, Future }`, right?

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/ParseFile.scala (hunk 2 above)
Comment (jkinkead) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/polyparser/ParseFile.scala]: Actually, I should say: Move all non-implicit imports. The execution context could stay here.

**Comment by jkinkead** [comment]
The concurrency looks fine. The only thing I'd say is to use the `global` execution context everywhere, since it's generally pretty good.

File: tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/RandomForest.scala (hunk 1 above)
Comment (mhrmm) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/RandomForest.scala]: Question: how do I specify the number of threads in the default execution context?

Comment (jkinkead) [inline tools/parse/src/main/scala/org/allenai/nlpstack/parse/poly/decisiontree/RandomForest.scala]: I don't know if you can. The initial size matches the number of processors you have. It's probably faster than specifying a thread pool - but if you're seeing better performance specifying something, document it here, and leave it as-is. :)

**Pull request merged**